### PR TITLE
Autodock vina tools

### PR DESCRIPTION
**Pull request by leobiscassi**
Hi @bgruening, I added my tools for docking protein-compounds according this thread [1] and this trello card [2].
It was told in [1], the docking tool not pass in planemo functional tests because the content of example output file isn't the same that output file, I need help with this. I tried install tool dependencies in test tool shed, but this was not totally validated.
I will be grateful if you can assist me with the functional tests on the docking tool and the final validation of the tool dependencies.

[1] [http://dev.list.galaxyproject.org/Galaxy-tools-for-docking-protein-compounds-td4668604.html](url)
[2] [https://trello.com/c/2eIu6SBa](url)

Best regards,
Leo Biscassi

### CODE REVIEW

File: chemicaltoolbox/datatypes/molecules.py [hunk 1]
```diff
@@ -449,6 +449,36 @@ def set_peek( self, dataset, is_multi_byte=False ):
             dataset.peek = 'file does not exist'
             dataset.blurb = 'file purged from disk'
 
+class PDBQT(GenericMolFile):
+    file_ext = "pdbqt"
+
```
Comment (bgruening) [inline chemicaltoolbox/datatypes/molecules.py]: Can you please add a link or short description as comment to the file type? This will serve as documentation. PDBqt it probably well known, but LOG I never heard about it :)

File: chemicaltoolbox/autodock_vina/prepare_receptor/prepare_receptor.xml [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    <help><![CDATA[
+        *** Tool for prepare receptor with Autodock Vina ***
+
+        For use this tool will be necessary perform the upload of a dataset, pdb file, and select it in receptor field.
+    ]]></help>
+    <citations>
+        <citation type="bibtex">
+            @misc{githubautodock_vina_tools,
+              author = {LastTODO, FirstTODO},
+              year = {TODO},
```
Comment (bgruening) [inline chemicaltoolbox/autodock_vina/prepare_receptor/prepare_receptor.xml]: TODO?

File: chemicaltoolbox/autodock_vina/prepare_receptor/prepare_receptor.xml [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        <data name="file_output" format="pdbqt" label="#echo os.path.splitext (str($receptor.name))[0]#.pdbqt" />
+    </outputs>
+    <tests>
+        <test>
+            <param name="receptor" value="3u1i_for_DM.pdb"/>
+            <output name="file_output" file="3u1i_for_DM.pdbqt"/>
+        </test>
+    </tests>
+    <help><![CDATA[
+        *** Tool for prepare receptor with Autodock Vina ***
```
Comment (bgruening) [inline chemicaltoolbox/autodock_vina/prepare_receptor/prepare_receptor.xml]: This could be extended a little bit ... `Tool to prepare a receptor (Protein) ...` maybe?

File: chemicaltoolbox/autodock_vina/prepare_receptor/prepare_receptor.xml [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        <exit_code range="1:" />
+    </stdio>
+    <description> - Tool for prepare receptor with Autodock Vina</description>
+    <command interpreter="python"><![CDATA[
+        prepare_receptor.py -r "$receptor" 
+                            -p "$file_output.files_path" 
+                            -f "$file_output"
+    ]]></command>
+    <inputs>
+        <param type="data" name="receptor" format="pdb" label="Receptor: " help="No data? See tip below" />
```
Comment (bgruening) [inline chemicaltoolbox/autodock_vina/prepare_receptor/prepare_receptor.xml]: `:` not needed ...

File: chemicaltoolbox/autodock_vina/prepare_receptor/prepare_receptor.xml [hunk 5]
```diff
@@ -0,0 +1,43 @@
+<tool id="prepare_receptor" name="Prepare receptor" version="0.1.0">
+    <requirements>
+        <requirement type="package" version="1.5.6">mgltools</requirement>
+    </requirements>
+    <stdio>
+        <exit_code range="1:" />
+    </stdio>
+    <description> - Tool for prepare receptor with Autodock Vina</description>
```
Comment (bgruening) [inline chemicaltoolbox/autodock_vina/prepare_receptor/prepare_receptor.xml]: `planemo lint` will complain here I guess. description needs to be before requirement.

File: chemicaltoolbox/autodock_vina/prepare_ligand/prepare_ligand.xml [hunk 6]
```diff
@@ -0,0 +1,43 @@
+<tool id="prepare_ligand" name="Prepare ligand" version="0.1.0">
+    <requirements>
+        <requirement type="package" version="1.5.6">mgltools</requirement>
+    </requirements>
+    <stdio>
+        <exit_code range="1:" />
+    </stdio>
+    <description> - Tool for prepare ligand with Autodock Vina</description>
+    <command interpreter="python"><![CDATA[
+        prepare_ligand.py -l "$ligand"
```
Comment (bgruening) [inline chemicaltoolbox/autodock_vina/prepare_ligand/prepare_ligand.xml]: the use of interpreter is deprecated and you should remove it.
Use `python $__tool_directory__/prepare_ligand.py` instead.

File: chemicaltoolbox/autodock_vina/prepare_ligand/prepare_ligand.xml [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+                          -p "$file_output.files_path"
+                          -f "$file_output"
+    ]]></command>
+    <inputs>
+        <param type="data" name="ligand" format="mol2" label="Ligand: " help="No data? See tip below" />
```
Comment (bgruening) [inline chemicaltoolbox/autodock_vina/prepare_ligand/prepare_ligand.xml]: `: ` not needed

File: chemicaltoolbox/autodock_vina/docking/test-data/NuBBE_1_obabel_3D_-_3u1i_for_DM.log [hunk 8]
```diff
@@ -0,0 +1,45 @@
+#################################################################
+# If you used AutoDock Vina in your work, please cite:          #
+#                                                               #
+# O. Trott, A. J. Olson,                                        #
+# AutoDock Vina: improving the speed and accuracy of docking    #
+# with a new scoring function, efficient optimization and       #
+# multithreading, Journal of Computational Chemistry 31 (2010)  #
+# 455-461                                                       #
+#                                                               #
+# DOI 10.1002/jcc.21334                                         #
```
Comment (bgruening) [inline chemicaltoolbox/autodock_vina/docking/test-data/NuBBE_1_obabel_3D_-_3u1i_for_DM.log]: This is the correct citation I guess, this should be included in the tool.

File: chemicaltoolbox/autodock_vina/docking/docking.xml [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+            <output name="file_output2" file="NuBBE_1_obabel_3D_-_3u1i_for_DM.log"/>
+        </test>
+    </tests>
+    <help><![CDATA[
+        *** Tool for perform docking with Autodock Vina ***
+
+        For use this tool will be necessary perform the upload of a datasets, pdbqt and text file's, and select it in yours fields.
+    ]]></help>
+    <citations>
+        <citation type="bibtex">
```
Comment (bgruening) [inline chemicaltoolbox/autodock_vina/docking/docking.xml]: `<citation type="doi">10.1002/jcc.21334</citation>`

File: chemicaltoolbox/autodock_vina/docking/docking.xml [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+                   -g "$file_output2"
+    ]]></command>
+    <inputs>
+        <param type="data" name="ligand" format="pdbqt" label="Ligands: " help="No data? See tip below" />
+        <param type="data" name="receptor" format="pdbqt" label="Receptor: " help="No data? See tip below" />
+        <param type="data" name="box" format="txt" label="Binding site coord.: " help="No data? See tip below" />
+    </inputs>
+    <outputs>
+        <data name="file_output1" format="pdbqt" label="#echo (os.path.splitext (str($ligand.name))[0]) + '_-_' + (os.path.splitext (str($receptor.name))[0])#.pdbqt" />
+        <data name="file_output2" format="log" label="#echo (os.path.splitext (str($ligand.name))[0]) + '_-_' + (os.path.splitext (str($receptor.name))[0])#.log" />
```
Comment (bgruening) [inline chemicaltoolbox/autodock_vina/docking/docking.xml]: Is this really a molecule format. From the test data I think this is just a normal text file. You can use `txt` for it and remove the additional file type for LOG.
I would actually not output it but pipe it to stdout. The advanced Galaxy user can see it there and we do not pollute the history with for every other user. What do you think?

File: chemicaltoolbox/autodock_vina/docking/docking.xml [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+                   -r "$receptor" 
+                   -b "$box" 
+                   -p "$file_output1.files_path" 
+                   -f "$file_output1" 
+                   -g "$file_output2"
+    ]]></command>
+    <inputs>
+        <param type="data" name="ligand" format="pdbqt" label="Ligands: " help="No data? See tip below" />
+        <param type="data" name="receptor" format="pdbqt" label="Receptor: " help="No data? See tip below" />
+        <param type="data" name="box" format="txt" label="Binding site coord.: " help="No data? See tip below" />
```
Comment (bgruening) [inline chemicaltoolbox/autodock_vina/docking/docking.xml]: This really needs to be documented, preferably in the help section.

File: chemicaltoolbox/autodock_vina/docking/docking.xml [hunk 12]
```diff
@@ -0,0 +1,52 @@
+<tool id="docking" name="Docking" version="0.1.0">
+    <requirements>
+        <requirement type="package" version="1.1.2">autodockvina</requirement>
```
Comment (bgruening) [inline chemicaltoolbox/autodock_vina/docking/docking.xml]: To trigger tests you need to create a shed.yml file ... it looks like that: https://github.com/bgruening/galaxytools/blob/master/tools/mcl/.shed.yml

**Comment by bgruening** [comment]
This looks really great. What we need to do is to create conda package for autodock this would actually replace the tool_dependecies.... conda is the new way and tool_depencies are the old way.

If you are not familiar with bioconda I can do this for you :)

File: chemicaltoolbox/autodock_vina/docking/docking.xml (hunk 12 above)
Comment (leobiscassi) [inline chemicaltoolbox/autodock_vina/docking/docking.xml]:
``` yml
name: package_autodockvina_1_1_2
owner: leobiscassi
description: Contains a tool dependency definition that downloads and extracts version 1.1.2 of autodock vina.
homepage_url: http://vina.scripps.edu/index.html
long_description: |
  AutoDock Vina is an open-source program for doing molecular docking. 
  It was designed and implemented by Dr. Oleg Trott in the Molecular Graphics Lab at The Scripps Research Institute.
type: tool_dependency_definition
categories:
  - Tool Dependency Packages
```
``` yml
name: package_mgltools_1_5_6
owner: leobiscassi
description: Contains a tool dependency definition that downloads and extracts version 1.5.6 of MGL Tools.
homepage_url: http://mgltools.scripps.edu/
long_description: |
  MGLTools is a software developed at the Molecular Graphics Laboratory (MGL)
  of The Scripps Research Institute for visualization and analysis of molecular structures. 
type: tool_dependency_definition
categories:
  - Tool Dependency Packages
```
I made this files, how can you see above, but I not included in this commit. In this case, where the tool dependencies file contains two packages, how I'll create the shed file? It would be how bellow?

``` yml
name: package_mgltools_1_5_6
owner: leobiscassi
description: Contains a tool dependency definition that downloads and extracts version 1.5.6 of MGL Tools.
homepage_url: http://mgltools.scripps.edu/
long_description: |
  MGLTools is a software developed at the Molecular Graphics Laboratory (MGL)
  of The Scripps Research Institute for visualization and analysis of molecular structures. 
type: tool_dependency_definition
categories:
  - Tool Dependency Packages

name: package_autodockvina_1_1_2
owner: leobiscassi
description: Contains a tool dependency definition that downloads and extracts version 1.1.2 of autodock vina.
homepage_url: http://vina.scripps.edu/index.html
long_description: |
  AutoDock Vina is an open-source program for doing molecular docking. 
  It was designed and implemented by Dr. Oleg Trott in the Molecular Graphics Lab at The Scripps Research Institute.
type: tool_dependency_definition
categories:
  - Tool Dependency Packages
```

File: chemicaltoolbox/autodock_vina/docking/docking.xml (hunk 11 above)
Comment (leobiscassi) [inline chemicaltoolbox/autodock_vina/docking/docking.xml]: I'll perform some improvements of documentation in the help section.

File: chemicaltoolbox/autodock_vina/docking/test-data/NuBBE_1_obabel_3D_-_3u1i_for_DM.log (hunk 8 above)
Comment (leobiscassi) [inline chemicaltoolbox/autodock_vina/docking/test-data/NuBBE_1_obabel_3D_-_3u1i_for_DM.log]: Included this citation in tool

File: chemicaltoolbox/datatypes/molecules.py (hunk 1 above)
Comment (leobiscassi) [inline chemicaltoolbox/datatypes/molecules.py]: Description was added and link too.

File: chemicaltoolbox/autodock_vina/prepare_receptor/prepare_receptor.xml (hunk 2 above)
Comment (leobiscassi) [inline chemicaltoolbox/autodock_vina/prepare_receptor/prepare_receptor.xml]: This problem was solved.

File: chemicaltoolbox/autodock_vina/prepare_ligand/prepare_ligand.xml (hunk 7 above)
Comment (leobiscassi) [inline chemicaltoolbox/autodock_vina/prepare_ligand/prepare_ligand.xml]: ok

File: chemicaltoolbox/autodock_vina/prepare_ligand/prepare_ligand.xml (hunk 6 above)
Comment (leobiscassi) [inline chemicaltoolbox/autodock_vina/prepare_ligand/prepare_ligand.xml]: ok

File: chemicaltoolbox/autodock_vina/prepare_receptor/prepare_receptor.xml (hunk 5 above)
Comment (leobiscassi) [inline chemicaltoolbox/autodock_vina/prepare_receptor/prepare_receptor.xml]: ok

File: chemicaltoolbox/autodock_vina/prepare_receptor/prepare_receptor.xml (hunk 4 above)
Comment (leobiscassi) [inline chemicaltoolbox/autodock_vina/prepare_receptor/prepare_receptor.xml]: ok

File: chemicaltoolbox/autodock_vina/prepare_receptor/prepare_receptor.xml (hunk 3 above)
Comment (leobiscassi) [inline chemicaltoolbox/autodock_vina/prepare_receptor/prepare_receptor.xml]: I'll perform some improvements of documentation in the help section.

File: chemicaltoolbox/autodock_vina/docking/docking.xml (hunk 9 above)
Comment (leobiscassi) [inline chemicaltoolbox/autodock_vina/docking/docking.xml]: ok

File: chemicaltoolbox/autodock_vina/docking/docking.xml (hunk 10 above)
Comment (leobiscassi) [inline chemicaltoolbox/autodock_vina/docking/docking.xml]: I removal log data type and we can use as simple text file, but I need this in tool output, because we use the informations about scoring that are stored in this file. More tools will be developed and they will need this information.
These tools are used for the supporting the [drugdesign](https://github.com/rodrigofaccioli/drugdesign) project, which being developed with [rodrigofaccioli.](https://github.com/rodrigofaccioli) We could talk more about our future plans if you think interesting.

**Comment by leobiscassi** [comment]
Hi @bgruening, can you review the changes that I did?

File: chemicaltoolbox/datatypes/molecules.py [hunk 13]
```diff
@@ -449,6 +449,33 @@ def set_peek( self, dataset, is_multi_byte=False ):
...
+    """
+    Autodock Format
+    The PDBQT format stores the atomic coordinates, partial charges and AutoDock atom types,
+    for both the receptor and the ligand.
+    http://autodock.scripps.edu/faqs-help/faq/what-is-the-format-of-a-pdbqt-file
+    """
+    file_ext = "pdbqt"
+
+    def sniff(self, filename):
+        if count_special_lines('COMPND', filename) > 0:
```
Comment (bgruening) [inline chemicaltoolbox/datatypes/molecules.py]: I'm not sure about the sniffer. As soon as a file contains COMPND or even worse ATOM this sniffer will return True. This is not intended, isn't it?
I guess you want to check for all these strings to be included in the file and if all occurrences it is a PDBQT file?

File: chemicaltoolbox/autodock_vina/docking/docking.py [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+    parser.add_argument("-g", action="store",
+                        dest="file_output2", required=True)
+
+    arguments = parser.parse_args()
+
+    return arguments
+
+
+def load_settings():
+    vina = os.environ["VINA"]
```
Comment (bgruening) [inline chemicaltoolbox/autodock_vina/docking/docking.py]: Why is this needed? You could simply assume that Vina is on your PATH isn't it?

File: chemicaltoolbox/autodock_vina/docking/docking.py [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+
+def convert_dat_for_pdbqt(path_dat_file, path_pdbqt):
+    if not os.path.isfile(path_dat_file):
+        raise EnvironmentError("*** ERROR: dataset not found! ***")
+
+    shutil.copy(path_dat_file, path_pdbqt)
+
+    for data_file in os.listdir(path_pdbqt):
+        if data_file.endswith(".dat"):
+            new_file = data_file.split(".")[0] + ".pdbqt"
```
Comment (bgruening) [inline chemicaltoolbox/autodock_vina/docking/docking.py]: Is this trick the main reason why you are using this extra python file?
You could do this all in your wrapper in cheetah code. For example like this one: `https://github.com/bgruening/galaxytools/blob/master/tools/diamond/diamond.xml#L10`

This would save you one file and all the argparse stuff, which is usually a huge overhead.

File: chemicaltoolbox/autodock_vina/docking/docking.py [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+
+    for file_name in os.listdir(path_pdbqt):
+        if file_name.endswith(".pdbqt"):
+            path_pdbqt_file = os.path.join(path_pdbqt, file_name)
+
+    return path_pdbqt_file
+
+
+def run_docking(autodock_vina, config_box, ligand, receptor, structure_file, log_file):
+    process = Popen([autodock_vina, '--config', config_box, '--receptor', receptor, '--ligand', ligand, '--out', structure_file, '--log', log_file], stdout=PIPE, stderr=PIPE)
```
Comment (bgruening) [inline chemicaltoolbox/autodock_vina/docking/docking.py]: This command can be called directly from cheetah.

**Comment by bgruening** [comment]
@leobiscassi thanks for the changes. I have a few additional comments. Please bear with me I'm trying to point out some things that can save you some time for future wrappers.
For example I think the python file can be replaced by cheetah code directly in your wrapper.

Moreover, I tried to build a conda package for Vina but failed to compile it by myself. This tool is not updated for 5 years as it seems and does not compile on recent computers/boost versions :(
I will work on this further.

**Comment by leobiscassi** [comment]
Hi @bgruening  ! Thanks for your support!
In this week I had a lot of other demands, I will take this job on weekend.
About your suggestion:

"Is this trick the main reason why you are using this extra python file?
You could do this all in your wrapper in cheetah code."

I will take the example and rewrite the wrapper without python file, thanks!

**Comment by bgruening** [comment]
@leobiscassi great! Don't care about the tool_dependency for now. I will create a conda recipe for it this will solve the problem hopefully.
Do you know if Vina is still maintained?

**Comment by bgruening** [comment]
@leobiscassi I have a conda package for autodock-vina. So installation should be very smooth. You can remove the tool_dependencies.xml file if you like.

xref: https://github.com/bioconda/bioconda-recipes/pull/1383

**Comment by leobiscassi** [comment]
Hi @bgruening , I did see the conda recipe, thanks a lot for help!
I don't have knowledge of how to works the installation of dependencies with the new way. Can you give me a example of this, please?
The installation will create the environment variable automatically?
Sorry about how I write and how express myself, I'm not a fluent english writer! :(

**Comment by bgruening** [comment]
@leobiscassi given that:
* you have a recent Galaxy instance
* conda-dependencies activated in `galaxy.ini`
* you tool `requirement` fits a package in the bioconda channel

Galaxy is able to automatically resolve the dependency for you. For example look at the travis configuration:
https://github.com/bgruening/galaxytools/blob/master/.travis.yml#L26
At first planemo will install the dependency for you and in the next step planemo will use this installation to test your tool against it.

For you this means if you change the `requirement` of your tool in this PR we will test your PR automatically for you :)

**Comment by leobiscassi** [comment]
@bgruening thank you! I'll try this.

**Comment by bgruening** [comment]
@leobiscassi have you seen this: http://jcheminf.springeropen.com/articles/10.1186/s13321-016-0139-1
Looks interesting, we could probably add this as two different modes.

**Comment by leobiscassi** [comment]
Hi @bgruening  , 
I've tried to create a conda package with the MGL Tools [1], another dependency of the tools, but I get the error message on the test step. I need run the script_ligand4 and script_receptor4 in the tools, this scripts are located in MGLToolsPckgs/AutoDockTools/Utilities24, the aliases script_ligand4 and script_receptor4 were created for this purpose, but not working on conda environment. I've tried add programs located on bin directory (pmv, adt and others), but without success either. Can you help me with this?
One more question, I'm developing a new tool for docking of protein-compound with FDA approval drugs library for drug repurposing, this library is one dependecy of this tool, can I create one conda package for this purpose?

[1] [https://github.com/leobiscassi/bioconda-recipes/tree/mgltools/recipes/mgltools](https://github.com/leobiscassi/bioconda-recipes/tree/mgltools/recipes/mgltools)

**Comment by bgruening** [comment]
@leobiscassi I commented on your commits in bioconda :)

>One more question, I'm developing a new tool for docking of protein-compound with FDA approval drugs library for drug repurposing, this library is one dependecy of this tool, can I create one conda package for this purpose?

This one I don't understand? A conda package is always good :)

**Comment by leobiscassi** [comment]
Hi @bgruening ,
Thank you for your support!
I'll see your comments and adjust the code. I asked about the conda package for library because it isn't a program and I don't knew if conda package is applicable in this situation. Thank you for your answer.

**Comment by bgruening** [comment]
@leobiscassi libraries are great additions for conda packages. You can pack everything with conda! For example we have packages for zlib, matplotlib .... So feel free to add those as well.

**Comment by leobiscassi** [comment]
@bgruening In this case is a library of mol2 files, yet makes sense for conda packages? So I'll create a new package for this library. Thanks a lot.

**Comment by bgruening** [comment]
@leobiscassi wait ... library in a sense of molecule library or programming library :)
Programming library is supported. Molecule library or any other kind of data storage is not :(

**Comment by leobiscassi** [comment]
@bgruening in this case is a molecule library. How can I solve this dependency with molecule library? Any suggestion?

**Comment by bgruening** [comment]
How big? Do you have a link? Is this drugbank? If so I would not include it as dependency. I would give the user the chance to include this library as sdf, smi ...
We have some nice tools in CTB to download drugbank or other libraries.

**Comment by leobiscassi** [comment]
This molecule library is a subset of the zinc database. [http://zinc.docking.org/] (http://zinc.docking.org/)

**Comment by bgruening** [comment]
Ah ZINC, yes even better. We have tools to download these kind of libraries. So define it as param="data" fomrat="sdf,smi,inchi,mol2,mol" and the user should download the latest verison before using your tool.

Does this makes sense to you?

**Comment by leobiscassi** [comment]
@bgruening Doesn't make sense in this case, because is the repurpose tool, the only subset what makes sense are the FDA approvals molecules, I would not give the user the option to select any subset in the ZINC database. Does make sense for you?

**Comment by bgruening** [comment]
Yes, but I would leave the decision up to the user. For example ZINC is only one database that can be used. DrugBank is an other (drug repositioning?). Or other tools like the CTB-fragmenter which supposedly creates new drug-like molecules.

I would add a huge hint for the user ... "please only use drug-like molecules like ZINC, DrugBank. You can check the drug likeness with our special drug-like tool if you are unsure."

And we don't have the problem of storing and offering the ZINC.

**Comment by leobiscassi** [comment]
humm ... ok, let's try it. Which is the tools for download data of these databases?

**Comment by bgruening** [comment]
These ones: https://github.com/bgruening/galaxytools/tree/master/chemicaltoolbox/data_source/get_online_data

**Comment by bgruening** [comment]
Btw. I'm working on a huge update: https://github.com/bgruening/galaxytools/pull/339

**Comment by leobiscassi** [comment]
I'm rewriting the tools of this PR for use only cheetah syntax as you suggested The unique dependency missing is the conda package of MGL Tools, but I'm testing in my local environment, thanks a lot for total support and suggestions! :smile:

**Comment by bgruening** [comment]
Awesome! I merged the MGLtools. Should be available in a few minutes. Let me know if you want to have merge your own PR on bioconda. I can add you.

**Comment by leobiscassi** [comment]
ok @bgruening thanks a lot! :smile:

**Comment by leobiscassi** [comment]
Hi @bgruening ! Can you check the new wrappers? :smile:

**Comment by bgruening** [comment]
Oh great! Can you please add such a file to this repo: https://github.com/bgruening/galaxytools/blob/master/chemicaltoolbox/confab/.shed.yml

This will enable testing of your tool with travis :)
I will have a look in a few hours.

**Comment by leobiscassi** [comment]
@bgruening one question, the section 'owner' still continues with the your user, right? :smile:

**Comment by bgruening** [comment]
Mh, yes ... we could think of having some cheminformatics group that we share in the TS. For the moment it's bgruening.

**Comment by leobiscassi** [comment]
@bgruening no problema! I'll add this file! thanks! :smile:

**Comment by leobiscassi** [comment]
@bgruening the file was added.

File: chemicaltoolbox/autodock_vina/.shed.yml [hunk 17]
```diff
@@ -0,0 +1,9 @@
+categories:
```
Comment (bgruening) [inline chemicaltoolbox/autodock_vina/.shed.yml]: This file needs to be either in every sub-folder or you can create suites like this one: https://github.com/galaxyproject/tools-iuc/tree/master/tools/datamash

A suite is more or less a combination of related tools that share a macro file and a common test-data folder. planemo understands by a slightly more complex .shed.yml (https://github.com/galaxyproject/tools-iuc/blob/master/tools/datamash/.shed.yml) how to create for every single tool one extra repo.
A suite has the big advantage of sharing test-data. I think you also have some pdbqt file that you can share between tools, isn't it.

It's up to you :)
If you decide to go for a suite, copy everything into one folder and enhance this one shed.yml file to the extended suite format.

File: chemicaltoolbox/datatypes/molecules.py (hunk 13 above)
Comment (bgruening) [inline chemicaltoolbox/datatypes/molecules.py]: @leobiscassi any comments here?

Comment (bgruening) [inline chemicaltoolbox/datatypes/molecules.py]: This datatype needs to go here: https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/datatypes/molecules.py
Galaxy now supports all chemical datatypes natively!! :)
I can do this is you like.

File: chemicaltoolbox/autodock_vina/prepare_receptor/prepare_receptor.xml [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+        <exit_code range="1" />
+    </stdio>
+    <command><![CDATA[
+        prepare_receptor4.py -r "$receptor" -o "$file_output" -v -U nphs_lps -A hydrogens
+    ]]></command>
+    <inputs>
+        <param type="data" name="receptor" format="pdb" label="Receptor: " help="Select a pdb file." />
+    </inputs>
+    <outputs>
+        <data name="file_output" format="pdbqt" label="#echo os.path.splitext (str($receptor.name))[0]#.pdbqt" />
```
Comment (bgruening) [inline chemicaltoolbox/autodock_vina/prepare_receptor/prepare_receptor.xml]: This somehow assumes that you input is labeld ****.pdb isn't it?
I don't think this will be often the case. I would take the standard Galaxy label.

**Comment by bgruening** [comment]
I have added a few questions. Please let me know if you will work in this otherwise I can try to help you and adjust things as you like.

**Comment by bgruening** [comment]
Btw. test are now working you can check travis logs why they are failing.

File: chemicaltoolbox/autodock_vina/prepare_receptor/prepare_receptor.xml (hunk 18 above)
Comment (leobiscassi) [inline chemicaltoolbox/autodock_vina/prepare_receptor/prepare_receptor.xml]: ```This somehow assumes that you input is labeld ****.pdb isn't it?```

yeah, I assumed this because I used this tool only with this file format in the case of receptors.

```I don't think this will be often the case. I would take the standard Galaxy label```

sorry, I don't understand, could you explain in detail? :disappointed_relieved:

**Comment by leobiscassi** [comment]
I added the .shed.yml files according to what you said me in the last commit.

File: chemicaltoolbox/autodock_vina/prepare_receptor/prepare_receptor.xml (hunk 18 above)
Comment (bgruening) [inline chemicaltoolbox/autodock_vina/prepare_receptor/prepare_receptor.xml]: Just remove the label="" :)

**Comment by bgruening** [comment]
One docking test is failing because autodock picks a random seed. Is there any way to specify the random-seed? This would be great for reproducibility.

**Comment by leobiscassi** [comment]
yes, I will fix this.

File: chemicaltoolbox/datatypes/molecules.py (hunk 13 above)
Comment (leobiscassi) [inline chemicaltoolbox/datatypes/molecules.py]: This file format is pretty similar to pdb file format, in cases I've been contact the head of file begins with one or other (COMPND or ATOM), I implemented this way because of it. Are you in agreement? If you have a better suggestion would be great! :smile: 
If you are in according I can do this work and commit in molecules.py.

```Galaxy now supports all chemical datatypes natively!! :)```

AWESOME!

Thanks for your help.

File: chemicaltoolbox/autodock_vina/prepare_receptor/prepare_receptor.xml (hunk 18 above)
Comment (leobiscassi) [inline chemicaltoolbox/autodock_vina/prepare_receptor/prepare_receptor.xml]: done!

**Comment by leobiscassi** [comment]
```yml

| Dataset Job Standard Output:
|  discarding /home/travis/conda/envs/iuc_conda/bin from PATH
|  prepending /tmp/tmpq8bvwP/job_working_directory/000/2/conda-env/bin to PATH
|  set verbose to  True
|  set outputfilename to  /tmp/tmpq8bvwP/files/000/dataset_2.dat
|  set cleanup to merge  nphs_lps
|  set repairs to  hydrogens
|  File Format unknown can't parse it
|  read  /tmp/tmpq8bvwP/files/000/dataset_1.dat
```

@bgruening In this case, the file (.dat) what was passed to script (prepare_ligand4.py) isn't a recognized type by script. Any suggestion? 
Early I had implemented the python script because of it. I feel what this will affect us too in the prepare receptor tool. :disappointed_relieved:

**Comment by bgruening** [comment]
@leobiscassi yeah this is a common obstacle for many tools that rely on file-endings. Try this, it's very easy:
```
ln -s $infile ./some_file_name.pdb &&
vina --input ./some_file_name.pdb
```

The idea is to link the file into the current working dir and give it a fixed name.

**Comment by leobiscassi** [comment]
hi @bgruening , thanks for the tip! tests now pass. about the pdbqt datatype, I responded in the last hours, see below:

> This file format is pretty similar to pdb file format, in cases I've been contact the head of file begins with one or other (COMPND or ATOM), I implemented this way because of it. Are you in agreement? If you have a better suggestion would be great! :smile: 
If you are in according I can do this work and commit in molecules.py.

> ```Galaxy  now supports all chemical datatypes natively!! :)```

> AWESOME!

> Thanks for your help.

File: chemicaltoolbox/datatypes/molecules.py (hunk 13 above)
Comment (bgruening) [inline chemicaltoolbox/datatypes/molecules.py]: My assumption is that this sniffer will also sniff PDB as PDBQT format.
Am I right that the last (non-empty line) needs to be `TORSDOF`? Also ROOT, ENDROOT and BRANCH is unique to PDBQT and do not appear in PDB isn't it?

If we integrate this format in Galaxy we can add a small test file as well and Galaxy will test the sniffer for us. This is definitely the way forward I would say.

**Comment by bgruening** [comment]
Oh man! This is beautiful! Test pass and we have autodock integration in CTB! How awesome is this!
 I will merge this now. You have waited far to long to get this in. I will fix some small things in an additional PR and ping you.

Let me know if you want to get the PDBQT format into Main Galaxy or if I should do this!
Thanks a bunch!

File: chemicaltoolbox/datatypes/molecules.py (hunk 13 above)
Comment (leobiscassi) [inline chemicaltoolbox/datatypes/molecules.py]: Yes, your assumptions are correct.

**Comment by leobiscassi** [comment]
@bgruening I would like to make changes if possible. :smile:

**Comment by bgruening** [comment]
I guess my PR with the other chemical datatypes should be a good guide: https://github.com/galaxyproject/galaxy/pull/1941
Thanks for working on this!

**Comment by bgruening** [comment]
Don't be afraid of the PR. It's just that large because I added a bunch of converters which you do not need.

**Comment by leobiscassi** [comment]
@bgruening Ok, I will work on this now. Thanks for the great orientation and all supporting. :grin:

**Comment by bgruening** [comment]
@leobiscassi no problem at all. Let's make Cheminformatics rock in Galaxy!

**Comment by leobiscassi** [comment]
@bgruening I made a few comments on my commit in galaxy molecules.py.

**Comment by leobiscassi** [comment]
@bgruening  last question, when the ChemicalToolBoX release with the autodock vina tools will be available? I'll pretty excited for testing this suite on my own galaxy instance! :grin:

**Comment by bgruening** [comment]
I uploaded it to the MTS. You should be able to install it. I need to work more on CTB in general to migrate everything to conda before I can do a new CTB release. But autodock you be installable.
If you can give me your TS usename I can add you to the autodock repos and you can update the repos as you like :)
I can also give you credentials to the other CTB repos if you are interested at some point.

**Comment by bgruening** [comment]
Thanks again!

**Comment by leobiscassi** [comment]
Hi @bgruening, my username in the main toolshed and test toolshed is leobiscassi, thanks! :smile:

**Comment by bgruening** [comment]
I added you to all three repos :)

**Comment by leobiscassi** [comment]
Thanks!
